lib/BuildSystem/LaneBasedExecutionQueue.cpp

```diff
@@ -281,28 +287,53 @@ class LaneBasedExecutionQueue : public BuildExecutionQueue {
     if (shouldCaptureOutput) {
       // Close the write end of the output pipe.
       ::close(outputPipe[1]);
-
-      // Read all the data from the output pipe.
-      while (true) {
-        char buf[4096];
-        ssize_t numBytes = read(outputPipe[0], buf, sizeof(buf));
-        if (numBytes < 0) {
+      ::close(stderrPipe[1]);
+
+      fd_set read_flags;
+      struct timeval waitd = {10, 0};
+      FD_ZERO(&read_flags);
+      FD_SET(stderrPipe[0], &read_flags);
+      FD_SET(outputPipe[0], &read_flags);
+      bool readStdout = false, readStderr = false;
+      while(true) {
+        int select_result = select(stderrPipe[0] + 1, &read_flags, nullptr, nullptr, &waitd);
+        if (select_result < 0) {
           getDelegate().commandProcessHadError(
               context.job.getForCommand(), handle,
-              Twine("unable to read process output (") + strerror(errno) + ")");
+              Twine("unable to use select() while reading from subprocess (") + strerror(errno) + ")");
           break;
         }
+        if (select_result == 0) break;
+
+        if (FD_ISSET(outputPipe[0], &read_flags)) {
+          FD_CLR(outputPipe[0], &read_flags);
+          if(!readFd(outputPipe[0], hasPartialOutput)) {
+            getDelegate().commandProcessHadError(
+                context.job.getForCommand(), handle,
+                Twine("unable to read process output (") + strerror(errno) + ")");
+            break;
+          }
+          readStdout = true;
+        }
```

Comment:
This isn't the right logic here... this will end up blocking reading all of the data off of the first FD which had data available, which bypasses the need for select in the first place. There are several reasons that won't work:
1. It doesn't ensure timely delivery of data for long running commands to clients.
2. But worse, it may cause deadlocks when the command writes to both stdout and stderr, but because data isn't being read from stderr the pipe fills up and the command itself blocks waiting to write. At that point, llbuild won't be able to finish reading from stdout (it hasn't closed) and the command won't be able to finish writing to it. This is the canonical reason that using `select()` is required in these circumstances.